src/type.c

```diff
@@ -75,6 +85,24 @@ static SEXP s3_type(SEXP x, struct vctrs_arg* x_arg) {
   return vec_slice(x, R_NilValue);
```

Comment:
It would be interesting to consider a version of `vec_slice()` that doesn't preserve names so we don't have to do as much work to unname after we slice. This feels somewhat similar to having a version of `vec_c()` that doesn't retain names

Reply:
If we did this, we'd have to reconsider how names are tracked in the lubridate PR. When we proxy, the names are added as a new column. This would get sliced and then restored even if we turned on the option to not preserve names.
https://github.com/tidyverse/lubridate/pull/871/files#diff-29798887c998d8e58008ea67fdf141d3R32

On the other hand, POSIXlt stores names directly on the `$year` field. So when we proxy we get a column that is named. If we turn off name preservation that should strip these names

Reply:
It seems like if we did this then we wouldn't even need a `vec_set_names()` call

Reply:
We may still need the set-names call on S3 objects (or at least on S3 data frames) because the names might be encoded in a field, as in POSIXlt vectors. Or maybe we should enforce names as a special `vctrs::rcrd_names` field in data frame proxies for performance and simplicity. Then vec-slice could rely on either the names attribute or the record field.

Reply:
> because the names might be encoded in a field, as in POSIXlt vectors

I think I'm arguing above that a `vec_slice(x, R_NilValue, preserve_names = false)` would already work correctly in this case.

Reply:
Maybe POSIXlt is not a good example. I was thinking the difference between named and unnamed record vectors of size zero would be the presence of a names field. This is equivalent to the presence or absence of a names attributes for normal vectors of size zero.